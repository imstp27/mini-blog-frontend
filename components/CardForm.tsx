import { Button, Form, Input, Select } from 'antd';
import React, { useEffect } from 'react';
import { MyCardProps } from './Card';

const { Option } = Select;

export interface CardFormProps {
  onSubmit: (values: any, isEdit: boolean) => void;
  editData: MyCardProps | null;
}

export const CardForm = (props: CardFormProps) => {
  const [form] = Form.useForm();

  const onFinish = (values: any) => {
    props.onSubmit(values, !!props.editData);
  };

  useEffect(() => {
    if (props.editData) {
      form.setFieldsValue({
        name: props.editData.name,
        status: props.editData.status,
        content: props.editData.content,
        category: props.editData.category,
      });
    } else form.resetFields();
  }, [props.editData]);

  return (
    <Form form={form} onFinish={onFinish}>
      <Form.Item name="name" rules={[{ required: true }]}>
        <Input placeholder="card name" />
      </Form.Item>
      <Form.Item name="status" rules={[{ required: true }]}>
        <Select placeholder="Select a status" allowClear>
          <Option value="new">New</Option>
          <Option value="publish">Publish</Option>
          <Option value="draft">Draft</Option>
          <Option value="banned">Banned</Option>
        </Select>
      </Form.Item>
      <Form.Item name="content" rules={[{ required: true }]}>
        <Input.TextArea placeholder="content" />
      </Form.Item>
      <Form.Item name="category" rules={[{ required: true }]}>
        <Select placeholder="Select a category" allowClear>
          <Option value="biology">Biology</Option>
          <Option value="finance">Finance</Option>
          <Option value="chemistry">Chemistry</Option>
          <Option value="engineering">Engineering</Option>
          <Option value="health">Health</Option>
          <Option value="sociology">Sociology</Option>
          <Option value="space">Space</Option>
          <Option value="art">Art</Option>
        </Select>
      </Form.Item>
      <Form.Item>
        <Button id="submit" type="primary" htmlType="submit" block>
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};
