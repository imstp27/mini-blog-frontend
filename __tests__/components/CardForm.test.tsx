import React from 'react'
import { CardForm, CardFormProps } from '@components/CardForm'
import { shallow, mount, ShallowWrapper } from 'enzyme'
import moment from 'moment';

const mock: CardFormProps = {
    editData: {
        _id: "id",
        name: "name",
        status: "new",
        category: "biology",
        content: "content",
        author: { _id: "user", name: "name", image: "image" },
        createdOn: moment().toISOString(),
        user: "user",
        onEdit: () => jest.fn(),
        onDelete: () => jest.fn()
    },
    onSubmit: () => jest.fn()
}

describe("CardForm component", () => {
    let wrapper: ShallowWrapper<CardFormProps>
    beforeAll(() => {
        wrapper = shallow(<CardForm {...mock} />);
    })
    it("should renders", () => {
        expect(wrapper.exists()).toBe(true);
    });
    it("should match snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });
});