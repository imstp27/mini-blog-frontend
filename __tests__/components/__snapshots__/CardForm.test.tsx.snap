// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardForm component should match snapshot 1`] = `
<ForwardRef(InternalForm)
  form={
    Object {
      "__INTERNAL__": Object {
        "itemRef": [Function],
      },
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "getInternalHooks": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "resetFields": [Function],
      "scrollToField": [Function],
      "setFields": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
    }
  }
  onFinish={[Function]}
>
  <FormItem
    name="name"
    rules={
      Array [
        Object {
          "required": true,
        },
      ]
    }
  >
    <Input
      placeholder="card name"
      type="text"
    />
  </FormItem>
  <FormItem
    name="status"
    rules={
      Array [
        Object {
          "required": true,
        },
      ]
    }
  >
    <ForwardRef(InternalSelect)
      allowClear={true}
      placeholder="Select a status"
    >
      <Option
        value="new"
      >
        New
      </Option>
      <Option
        value="publish"
      >
        Publish
      </Option>
      <Option
        value="draft"
      >
        Draft
      </Option>
      <Option
        value="banned"
      >
        Banned
      </Option>
    </ForwardRef(InternalSelect)>
  </FormItem>
  <FormItem
    name="content"
    rules={
      Array [
        Object {
          "required": true,
        },
      ]
    }
  >
    <ForwardRef
      placeholder="content"
    />
  </FormItem>
  <FormItem
    name="category"
    rules={
      Array [
        Object {
          "required": true,
        },
      ]
    }
  >
    <ForwardRef(InternalSelect)
      allowClear={true}
      placeholder="Select a category"
    >
      <Option
        value="biology"
      >
        Biology
      </Option>
      <Option
        value="finance"
      >
        Finance
      </Option>
      <Option
        value="chemistry"
      >
        Chemistry
      </Option>
      <Option
        value="engineering"
      >
        Engineering
      </Option>
      <Option
        value="health"
      >
        Health
      </Option>
      <Option
        value="sociology"
      >
        Sociology
      </Option>
      <Option
        value="space"
      >
        Space
      </Option>
      <Option
        value="art"
      >
        Art
      </Option>
    </ForwardRef(InternalSelect)>
  </FormItem>
  <FormItem>
    <Button
      block={true}
      htmlType="submit"
      id="submit"
      type="primary"
    >
      Submit
    </Button>
  </FormItem>
</ForwardRef(InternalForm)>
`;
